File: KNAP.cpp
/*
---------------luyencode.net----------------------------------
---------------Problem: KNAP-------------------------------
---------------Author: HeHe---------------------------------
---------------Time: 2024-10-19 22:52:53.959096+07:00---------
*/
#include <iostream>
#include <iomanip>
#include <vector>
#include <queue>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <assert.h>
#include <algorithm>
#include <functional>
#include <numeric>
#include <math.h>
#include <cstdint>
#include <climits>

#define FAST_IO ios::sync_with_stdio(false);cin.tie(0);cout.tie(0);
using namespace std;

const int MOD = 1000000007;

int dp[5001];
int items[50001][2];

int main() {
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    FAST_IO;
    // --------------- Solve -------------------
    int n, M;
    cin >> n >> M;

    for (int i = 0; i < n; ++i) {
        cin >> items[i][0] >> items[i][1];
    }
    int ans = 0;
    for (int i = 0; i < n; ++i) {
        for (int j = M; j >= items[i][0]; --j) {
            dp[j] = max(dp[j], items[i][1] + dp[j - items[i][0]]);
            ans = max(ans, dp[j]);
        }
    }
    cout << ans;
    ////////////////////////////////////////////
    return 0;
}

File: KNAP.py
"""
---------------luyencode.net----------------------------------
---------------Problem: KNAP-------------------------------
---------------Author: HeHe---------------------------------
---------------Time: 2024-10-19 22:48:56.672357+07:00---------
"""
in1 = lambda: int(input())
in2 = lambda: list(map(int, input().split()))
in3 = lambda: map(int, input().split())
MOD = 1000000007
# --------------- Solve ---------------  # 
n, M = in3()
arr = []
for i in range(n):
    arr[i][0], arr[i][1] = in3()
    arr.append(arr[i][0], arr[i][1])

dp = [0] * (M + 1)
ans = 0
for i in range(n):
    for j in range(M, arr[i][0] - 1, -1):
        dp[j] = max(dp[j], arr[i][1] + dp[j - arr[i][0]])
        ans = max(ans, dp[j])

print(ans)
##########################################
